contrib/drivers/netapp/ontap/ontap.go

```diff
@@ -0,0 +1,463 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package ontap
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"github.com/ghodss/yaml"
+	log "github.com/golang/glog"
+	uuid "github.com/satori/go.uuid"
+	"strconv"
+	"strings"
+
+	"github.com/netapp/trident/storage"
+	sa "github.com/netapp/trident/storage_attribute"
+	drivers "github.com/netapp/trident/storage_drivers"
+	"github.com/netapp/trident/storage_drivers/ontap"
+	"github.com/netapp/trident/storage_drivers/ontap/api"
+	"github.com/netapp/trident/utils"
+
+	odu "github.com/opensds/opensds/contrib/drivers/utils"
+	. "github.com/opensds/opensds/contrib/drivers/utils/config"
+	"github.com/opensds/opensds/pkg/model"
+	pb "github.com/opensds/opensds/pkg/model/proto"
+	"github.com/opensds/opensds/pkg/utils/config"
+)
+
+func lunPath(name string) string {
+	return fmt.Sprintf("/vol/%v/lun0", name)
+}
+
+func getVolumeName(id string) string {
+	r := strings.NewReplacer("-", "")
+	return volumePrefix + r.Replace(id)
+}
+
+func getSnapshotName(id string) string {
+	r := strings.NewReplacer("-", "")
+	return snapshotPrefix + r.Replace(id)
+}
+
+// Get LUN Serial Number
+func (d *Driver) getLunSerialNumber(lunPath string) (string, error) {
+
+	lunSrNumber, err := d.sanStorageDriver.API.LunGetSerialNumber(lunPath)
+	if err != nil {
+		return "", fmt.Errorf("problem reading maps for LUN %s: %v", lunPath, err)
+	}
+
+	return naaPrefix + hex.EncodeToString([]byte(lunSrNumber.Result.SerialNumber())), nil
+}
+
+func (d *Driver) GetVolumeConfig(name string, size int64) (volConfig *storage.VolumeConfig) {
+	volConfig = &storage.VolumeConfig{
+		Version:      VolumeVersion,
+		Name:         name,
+		InternalName: name,
+		Size:         strconv.FormatInt(size*bytesGB, 10),
+		Protocol:     d.sanStorageDriver.GetProtocol(),
+		AccessMode:   accessMode,
+		VolumeMode:   volumeMode,
+		AccessInfo:   utils.VolumeAccessInfo{},
+	}
+	return volConfig
+}
+
+func (d *Driver) GetSnapshotConfig(snapName string, volName string) (snapConfig *storage.SnapshotConfig) {
+	snapConfig = &storage.SnapshotConfig{
+		Version:            SnapshotVersion,
+		Name:               snapName,
+		InternalName:       snapName,
+		VolumeName:         volName,
+		VolumeInternalName: volName,
+	}
+	return snapConfig
+}
+
+func (d *Driver) Setup() error {
+	// Read NetApp ONTAP config file
+	d.conf = &ONTAPConfig{}
+
+	p := config.CONF.OsdsDock.Backends.NetappOntapSan.ConfigPath
+	if "" == p {
+		p = defaultConfPath
+	}
+	if _, err := Parse(d.conf, p); err != nil {
+		return err
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("Unable to instantiate backend")
+		}
```

Comment:
please follow https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
Done, Thank You